text/0000-pmgr.md

```diff
@@ -0,0 +1,257 @@
+happen at byte-level granularity, however it's an error to submit a page request that is non-page-aligned. What does
+byte-level granularity eviction mean? It means that region needs to reset to stable-storage, including bytes that may
+not cover a full page.
+
+When the kernel requests information about a page, it sends a set of `ObjectRange` structs for a given object ID. The
+pager responds with a set of `PhysRange` structs. The pager's response does not need to fulfill the entire request, as
+long as at least one page is filled (probably the first one). The kernel reads out physical pages, in order as specified
+by the completion, and maps them in sequence into the object's page tree.
+
+The pager, at any time, can prefetch (if it has memory) and inform the kernel of its prefetching work. It does this via
+the tuple `(PhysRange, ObjectRange)` in the `PageInfo` variant of the `PagerRequest` enum.
+
+# Drawbacks
+[drawbacks]: #drawbacks
+
+Ultimately, the system needs a manager of stable storage and a mechanism to move object memory
+safely to and from that storage.
+
+# Rationale and alternatives
+[rationale-and-alternatives]: #rationale-and-alternatives
+
+The pager is designed to fit into userspace, because it cannot be easily implemented in-kernel. This
+is because the drivers for devices that contain stable storage will be implemented in userspace, so
+the pager (which sits above them) should be as well. As a result, most of the design decisions are
+in-place to avoid the overheads of the kernel boundary crossings (escrow for DRAM pages, hinting,
+etc.). The choice to bypass the object system for NVM is done for this same reason.
+
+Another major component of this is the idea that a user application could submit syncing requests to
+the pager in a more fine-grained way than "sync(2)" or "sync this object". The hope is to give
+enough control to a transaction manager that could then efficiently sync only the pages that it
+needs to to ensure failure-atomicity.
+
+However, syncing via the kernel is slow. Ideally, userspace would be able to request the pager sync without involving
+the kernel at all. We plan to do this, see Future Possibilities. However, even with that system in-place, we want the
+kernel to be able to submit sync requests. This is because it probably needs this support anyway, but we also need to
+allow the kernel to send Sync requests to the pager on behalf of applications that may not be able to communicate with
+the pager directly for some reason.
+
+# Prior art
+[prior-art]: #prior-art
+
+This approach is similar to standard microkernel designs where the functionality of paging and
+drivers for stable storage cannot be added to the kernel.
+
+# Unresolved questions
+[unresolved-questions]: #unresolved-questions
+
+- What should the values of the constants A-I, above, be?
+- What should the format of the on-stable-storage data be?
```

Comment:
I could see there being a need for multiple stable storage formats. The existing object format could work, but for byte-granular `Evict` requests, this would be too much overhead. For instance, we don't need the FOT since it exists somewhere else. I manage a persitant memory allocator-like mechanism for evictions in the stable storage manager.

Reply:
I'm talking mostly about the storage format for organizing the KV-like interface for getting object data, not object data itself (which should be stored unchanged)